art.hpp

```diff
@@ -47,26 +81,314 @@ class db final {
+
+    // Iff the iterator is positioned on an index entry, then returns
+    // the decoded key associated with that index entry.
+    std::optional<const key> get_key() noexcept;
+
+    // Iff the iterator is positioned on an index entry, then returns
+    // the value associated with that index entry.
+    std::optional<const value_view> get_val() const noexcept;
+
+    bool operator==(const iterator& other) const noexcept;
+    bool operator!=(const iterator& other) const noexcept;
+
+    // Debugging
+    [[gnu::cold]] UNODB_DETAIL_NOINLINE void dump(std::ostream& os) const;
+
+   protected:
+    // Push the given node onto the stack and traverse from the
+    // caller's node to the left-most leaf under that node, pushing
+    // nodes onto the stack as they are visited.
+    iterator& left_most_traversal(detail::node_ptr node) noexcept;
+
+    // Descend from the current state of the stack to the right most
+    // child leaf, updating the state of the iterator during the
+    // descent.
+    iterator& right_most_traversal(detail::node_ptr node) noexcept;
+
+    // Compare the given key (e.g., the to_key) to the current key in
+    // the internal buffer.
+    //
+    // @return -1, 0, or 1 if this key is LT, EQ, or GT the other key.
+    int cmp(const detail::art_key& akey) const noexcept {
+      // TODO(thompsonbry) Explore a cheaper way to handle the
+      // exclusive bound case when developing variable length key
+      // support based on the maintained key buffer.
+      UNODB_DETAIL_ASSERT(!stack_.empty());
+      auto& node = stack_.top().node;
+      UNODB_DETAIL_ASSERT(node.type() == node_type::LEAF);
+      const auto* const leaf{node.ptr<detail::leaf*>()};
+      return leaf->get_key().cmp(akey);
+    }
+
+    //
+    // stack access methods.
+    //
+
+    // Return true unless the stack is empty.
+    bool valid() const noexcept { return !stack_.empty(); }
+
+    // Return true iff the stack is empty.
+    bool empty() const noexcept { return stack_.empty(); }
+
+    // Push an entry onto the stack.
+    void push(const stack_entry& e) noexcept { stack_.push(e); }
+
+    // Push an entry onto the stack.
+    //
+    // TODO(thompsonbry) handle variable length keys here.
+    void push(detail::node_ptr node, std::byte key_byte,
+              std::uint8_t child_index) noexcept {
+      stack_.push({node, key_byte, child_index});
+    }
+
+    // Push a leaf onto the stack.
+    void push_leaf(detail::node_ptr aleaf) noexcept {
+      // Mock up a stack entry for the leaf. The [key] and
+      // [child_index] are ignored for a leaf.
+      push(aleaf, static_cast<std::byte>(0xFFU),
+           static_cast<std::uint8_t>(0xFFU));
+    }
+
+    // Pop an entry from the stack.
+    //
+    // TODO(thompsonbry) handle variable length keys here.
+    void pop() noexcept { stack_.pop(); }
+
+    // Return the entry (if any) on the top of the stack.
+    auto top() noexcept { return stack_.top(); }
```

Comment:
Can you spell out the `auto` type here? The existing usage of `auto` return types, except for template-mandated cases, looks like a mistake to me
```suggestion
    [[nodiscard]] auto top() noexcept { return stack_.top(); }
```